debian/patches/91_sulogin_lockedpw.dpatch (Thom May):
Purpose: Make sure file systems can be fixed on machines with locked
         root accounts too, by presenting a shell in these cases.
Authour: Thom May and Ubuntu.
Fixes:   #326678
Status:  unknown

--- a/src/sulogin.c
+++ b/src/sulogin.c
@@ -329,7 +329,11 @@ struct passwd *getrootpwent(int try_manu
 		fprintf(stderr, "%s: no entry for root\n", F_SHADOW);
 		strcpy(pwd.pw_passwd, "");
 	}
-	if (!valid(pwd.pw_passwd)) {
+
+	/* disabled passwords are valid too */
+	if (!(strcmp(pwd.pw_passwd, "*") == 0) ||
+	    !(strcmp(pwd.pw_passwd, "!") == 0) ||
+	    !valid(pwd.pw_passwd)) {
 		fprintf(stderr, "%s: root password garbled\n", F_SHADOW);
 		strcpy(pwd.pw_passwd, ""); }
 	return &pwd;
@@ -585,6 +589,14 @@ int main(int argc, char **argv)
 		fprintf(stderr, "sulogin: cannot open password database!\n");
 		sleep(2);
 	}
+	/*
+	 *	If the root password is locked, fire up a shell
+	 */
+	if ((strcmp(pwd->pw_passwd, "*") == 0) ||
+	    (strcmp(pwd->pw_passwd, "!") == 0)) {
+		fprintf(stderr, "sulogin: root account is locked, starting shell\n");
+		sushell(pwd);
+	}
 
 	/*
 	 *	Ask for the password.
--- a/man/sulogin.8
+++ b/man/sulogin.8
@@ -39,6 +39,9 @@ Give root password for system maintenanc
 .br
 (or type Control\-D for normal startup):
 .PP
+If the root account is locked, no password prompt is displayed and
+\fIsulogin\fR behaves as if the correct password were entered.
+.PP
 \fIsulogin\fP will be connected to the current terminal, or to the
 optional device that can be specified on the command line
 (typically \fB/dev/console\fP).
